### PR TITLE
Replace size/weight estimate tuple with struct for named fields

### PR DESCRIPTION
**Pull request by instagibbs**
For clarity of return values of size estimation functions.

### CODE REVIEW

**Review by jonatack — COMMENTED** [review]
ACK 79e2a276e4817a25c8c8a757dc6ce8e52c1b37f6

Good idea; the code is much clearer with the struct.  It might be nice to clang-format the touched function signatures.

**Review by theStack — APPROVED** [review]
Code review ACK 79e2a276e4817a25c8c8a757dc6ce8e52c1b37f6

**Comment by achow101** [comment]
ACK 79e2a276e4817a25c8c8a757dc6ce8e52c1b37f6

**Review by ryanofsky — APPROVED** [review]
Code review ACK 79e2a276e4817a25c8c8a757dc6ce8e52c1b37f6. I was going to suggest returning `std::optional<TxSize>` instead of `TxSize` to get rid of magic `return TxSize{-1, -1};` negative sizes, but this would be a bigger change and -1 seems to be used outside of this anyway.

**Comment by practicalswift** [comment]
Concept ACK

Non-blocking suggestion: Could in-class initialize `vsize` and `weight` to reasonable defaults?

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -1390,8 +1390,13 @@ class WalletRescanReserver
 * Use DummySignatureCreator, which inserts 71 byte signatures everywhere.
 * NOTE: this requires that all inputs must be in mapWallet (eg the tx should
 * be IsAllFromMe). */
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, bool use_max_sig = false) EXCLUSIVE_LOCKS_REQUIRED(wallet->cs_wallet);
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, const std::vector<CTxOut>& txouts, bool use_max_sig = false);
+struct TxSize {
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:1393]: The doxygen comment is attached wrong

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -1390,8 +1390,13 @@ class WalletRescanReserver
...
+    int64_t vsize;
+    int64_t weight;
+};
+
+TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, bool use_max_sig = false) EXCLUSIVE_LOCKS_REQUIRED(wallet->cs_wallet);
+TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, const std::vector<CTxOut>& txouts, bool use_max_sig = false);
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:1399]: could clang-format-diff?

File: src/wallet/feebumper.cpp [hunk 3]
```diff
@@ -190,7 +190,7 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
     if (coin_control.m_feerate) {
         // The user provided a feeRate argument.
         // We calculate this here to avoid compiler warning on the cs_wallet lock
-        const int64_t maxTxSize = CalculateMaximumSignedTxSize(*wtx.tx, &wallet).first;
+        const int64_t maxTxSize = CalculateMaximumSignedTxSize(*wtx.tx, &wallet).vsize;
```
Comment (MarcoFalke) [inline src/wallet/feebumper.cpp:193]:
```suggestion
        const int64_t maxTxSize{CalculateMaximumSignedTxSize(*wtx.tx, &wallet).vsize};
```

**Review by maflcko — COMMENTED** [review]
left some nits and pointed out that doxygen will be messed up

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -1390,8 +1390,13 @@ class WalletRescanReserver
 * Use DummySignatureCreator, which inserts 71 byte signatures everywhere.
 * NOTE: this requires that all inputs must be in mapWallet (eg the tx should
 * be IsAllFromMe). */
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, bool use_max_sig = false) EXCLUSIVE_LOCKS_REQUIRED(wallet->cs_wallet);
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, const std::vector<CTxOut>& txouts, bool use_max_sig = false);
+struct TxSize {
+    int64_t vsize;
+    int64_t weight;
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:1395]:
```suggestion
    const int64_t vsize{-1};
    const int64_t weight{-1};
```

File: src/wallet/wallet.h (hunk 1 above)
Comment (instagibbs) [reply src/wallet/wallet.h]: fixed(I think)

File: src/wallet/wallet.h [hunk 5]
```diff
@@ -1386,12 +1386,17 @@ class WalletRescanReserver
...
+};
+
 /** Calculate the size of the transaction assuming all signatures are max size
 * Use DummySignatureCreator, which inserts 71 byte signatures everywhere.
 * NOTE: this requires that all inputs must be in mapWallet (eg the tx should
 * be IsAllFromMe). */
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, bool use_max_sig = false) EXCLUSIVE_LOCKS_REQUIRED(wallet->cs_wallet);
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, const std::vector<CTxOut>& txouts, bool use_max_sig = false);
+TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, bool use_max_sig = false) EXCLUSIVE_LOCKS_REQUIRED(wallet->cs_wallet);
+TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, const std::vector<CTxOut>& txouts, bool use_max_sig = false);
```
Comment (MarcoFalke) [inline src/wallet/wallet.h:1399]:
```suggestion
TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* wallet, const std::vector<CTxOut>& txouts, bool use_max_sig = false);
```

;)

**Review by maflcko — APPROVED** [review]
ACK

File: src/wallet/wallet.h (hunk 5 above)
Comment (instagibbs) [reply src/wallet/wallet.h:1399]: hmmm didn't see any output running locally. Probably using it wrong :)

File: src/wallet/wallet.h (hunk 4 above)
Comment (instagibbs) [reply src/wallet/wallet.h]: const makes it unhappy due to default copy constructor being ill-formed, removed

**Review by maflcko — APPROVED** [review]
ACK

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -1628,14 +1628,14 @@ bool CWallet::ImportScriptPubKeys(const std::string& label, const std::set<CScri
 }
 
 // Returns pair of vsize and weight
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, bool use_max_sig)
+TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, bool use_max_sig)
```
Comment (jonatack) [inline src/wallet/wallet.cpp:1631]: clang-format

File: src/wallet/wallet.cpp [hunk 7]
```diff
@@ -1644,16 +1644,16 @@ std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx,
 }
 
 // txouts needs to be in the order of tx.vin
-std::pair<int64_t, int64_t> CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, const std::vector<CTxOut>& txouts, bool use_max_sig)
+TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *wallet, const std::vector<CTxOut>& txouts, bool use_max_sig)
```
Comment (jonatack) [inline src/wallet/wallet.cpp:1647]: same

**Comment by jonatack** [comment]
utACK 299e85c8188510528f1e49a0b8d2a99f45916898

good changes in `git diff 79e2a27 299e85c`, modulo clang format nits

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -1628,14 +1628,14 @@ bool CWallet::ImportScriptPubKeys(const std::string& label, const std::set<CScri
 }
 
 // Returns pair of vsize and weight
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp:1630]: Can remove this comment?

Comment (instagibbs) [reply src/wallet/wallet.cpp:1630]: removed

**Comment by instagibbs** [comment]
rebased on master

**Comment by MarcoFalke** [comment]
review ACK 881a3e2e17c5a6fdffb16a47a2eaff9029f261e5

**Comment by practicalswift** [comment]
cr ACK 881a3e2e17c5a6fdffb16a47a2eaff9029f261e5

Much more readable!

Possible candidates for similar treatment (out of scope for this PR of course):

```
$ git grep -E '^( |std::optional<)*std::pair.*\(' -- "*.h"
src/consensus/tx_verify.h:std::pair<int, int64_t> CalculateSequenceLocks(const CTransaction &tx, int flags, std::vector<int>& prevHeights, const CBlockIndex& block);
src/httpserver.h:    std::pair<bool, std::string> GetHeader(const std::string& hdr) const;
src/indirectmap.h:    std::pair<iterator, bool> insert(const value_type& value) { return m.insert(value); }
src/pubkey.h:    std::optional<std::pair<XOnlyPubKey, bool>> CreateTapTweak(const uint256* merkle_root) const;
src/rpc/util.h:std::pair<int64_t, int64_t> ParseDescriptorRange(const UniValue& value);
src/txorphanage.h:    std::pair<CTransactionRef, NodeId> GetTx(const uint256& txid) const EXCLUSIVE_LOCKS_REQUIRED(g_cs_orphans);
src/util/readwritefile.h:std::pair<bool,std::string> ReadBinaryFile(const fs::path &filename, size_t maxsize=std::numeric_limits<size_t>::max());
src/util/system.h:    std::pair<int, char**> get();
```

**Review by promag — COMMENTED** [review]
Code review ACK 881a3e2e17c5a6fdffb16a47a2eaff9029f261e5.

**Pull request merged**